8.0.1 compatibility

Would be nice to have 8.0.1 compatibility for this package.

Presuming this is the same issue; a fresh install of slack fails on installing present.

```
--  While building package present-4.0.2 using:
      /Users/jowens/.stack/setup-exe-cache/x86_64-osx/Cabal-simple_mPHDZzAJ_1.24.0.0_ghc-8.0.1 --builddir=.stack-work/dist/x86_64-osx/Cabal-1.24.0.0 build --ghc-options " -ddump-hi -ddump-to-file"
    Process exited with code: ExitFailure 1
    Logs have been written to: /Users/jowens/.stack/global-project/.stack-work/logs/present-4.0.2.log

    Configuring present-4.0.2...
    Building present-4.0.2...
    Preprocessing library present-4.0.2...
    [1 of 2] Compiling Control.Monad.Trans.State.Strict ( src/Control/Monad/Trans/State/Strict.hs, .stack-work/dist/x86_64-osx/Cabal-1.24.0.0/build/Control/Monad/Trans/State/Strict.o )
    [2 of 2] Compiling Present          ( src/Present.hs, .stack-work/dist/x86_64-osx/Cabal-1.24.0.0/build/Present.o )

    /private/var/folders/wn/2ppvrfkd7g18c3k5gsjq6b900000gn/T/stack12444/present-4.0.2/src/Present.hs:303:18: error:
        â¢ The constructor âDataDâ should have 6 arguments, but has been given 5
        â¢ In the pattern: DataD _cxt _ vars cons _deriving
          In a case alternative:
              DataD _cxt _ vars cons _deriving
                -> do { cs <- mapM makeConstructor cons;
                        return
                          (Just
                             (DataTypeDefinition
                                typeConstructor (DataType (map toTypeVariable vars) cs))) }
          In the expression:
            case dec of {
              DataD _cxt _ vars cons _deriving
                -> do { cs <- mapM makeConstructor cons;
                        return
                          (Just
                             (DataTypeDefinition
                                typeConstructor (DataType (map toTypeVariable vars) cs))) }
              NewtypeD _cxt _ vars con _deriving
                -> do { c <- makeConstructor con;
                        return
                          (Just
                             (DataTypeDefinition
                                typeConstructor (DataType (map toTypeVariable vars) [...]))) }
              TySynD _ vars ty
                -> do { ty' <- normalizeType ty;
                        return
                          (Just
                             (TypeAliasDefinition
                                typeConstructor (TypeAlias (map toTypeVariable vars) ty'))) }
              _ -> fail "Not a supported data type declaration." }

    /private/var/folders/wn/2ppvrfkd7g18c3k5gsjq6b900000gn/T/stack12444/present-4.0.2/src/Present.hs:952:40: error:
        â¢ The constructor âInstanceDâ should have 4 arguments, but has been given 3
        â¢ In the pattern:
            InstanceD _ (AppT (ConT _className) (ConT typeName)) _
          In a case alternative:
              InstanceD _ (AppT (ConT _className) (ConT typeName)) _
                -> Just (TypeConstructor typeName, ValueVariable method)
          In the expression:
            case i of {
              InstanceD _ (AppT (ConT _className) (ConT typeName)) _
                -> Just (TypeConstructor typeName, ValueVariable method)
              _ -> Nothing }

    /private/var/folders/wn/2ppvrfkd7g18c3k5gsjq6b900000gn/T/stack12444/present-4.0.2/src/Present.hs:1025:14: error:
        â¢ The constructor âVarIâ should have 3 arguments, but has been given 4
        â¢ In the pattern: VarI _ ty _ _
          In the pattern: Just (VarI _ ty _ _)
          In a case alternative:
              Just (VarI _ ty _ _) -> makeTypePresenter (return ty)
```

--- a/.travis.yml
+++ b/.travis.yml
@@ -1,0 +1,27 @@
+language: generic
+
+cache:
+  directories:
+  - $HOME/.stack
+
+addons:
+  apt:
+    packages:
+      - libgmp-dev
+
+matrix:
+  include:
+  - env: STACK_YAML=lts-6.yaml
+  - env: STACK_YAML=lts-7.yaml
+  - env: STACK_YAML=stack.yaml
+
+before_install:
+- mkdir -p ~/.local/bin
+- export PATH=$HOME/.local/bin:$PATH
+- travis_retry curl -L https://www.stackage.org/stack/linux-x86_64 | tar xz --wildcards --strip-components=1 -C ~/.local/bin '*/stack'
+
+install:
+- stack --no-terminal --install-ghc test --only-dependencies
+
+script:
+- stack --no-terminal test --haddock --no-haddock-deps

--- a/lts-6.yaml
+++ b/lts-6.yaml
@@ -1,0 +1,3 @@
+packages:
+- '.'
+resolver: lts-6.30

--- a/lts-7.yaml
+++ b/lts-7.yaml
@@ -1,0 +1,3 @@
+packages:
+- '.'
+resolver: lts-7.20

--- a/src/Present.hs
+++ b/src/Present.hs
@@ -121,24 +121,24 @@ normalizeType = go
                          else NormalCons (TypeConstructor name))
             TH.TupleT i ->
               case lookup i tupleConstructors of
-                Nothing -> fail ("Tuple arity " ++ show i ++ " not supported.")
+                Nothing -> Left ("Tuple arity " ++ show i ++ " not supported.")
                 Just cons -> return (NormalCons (TypeConstructor cons))
             TH.ListT -> return (NormalCons (TypeConstructor ''[]))
-            TH.PromotedT _ -> fail "Promoted types are not supported."
-            TH.UnboxedTupleT _ -> fail "Unboxed tuples are not supported."
-            TH.ArrowT -> fail "The function arrow (->) is not supported."
-            TH.EqualityT -> fail "Equalities are not supported."
-            TH.PromotedTupleT _ -> fail "Promoted types are not supported."
-            TH.PromotedNilT -> fail "Promoted types are not supported."
-            TH.PromotedConsT -> fail "Promoted types are not supported."
-            TH.StarT -> fail "Star (*) is not supported."
-            TH.ConstraintT -> fail "Constraints are not supported."
-            TH.LitT _ -> fail "Type-level literals are not supported."
+            TH.PromotedT _ -> Left "Promoted types are not supported."
+            TH.UnboxedTupleT _ -> Left "Unboxed tuples are not supported."
+            TH.ArrowT -> Left "The function arrow (->) is not supported."
+            TH.EqualityT -> Left "Equalities are not supported."
+            TH.PromotedTupleT _ -> Left "Promoted types are not supported."
+            TH.PromotedNilT -> Left "Promoted types are not supported."
+            TH.PromotedConsT -> Left "Promoted types are not supported."
+            TH.StarT -> Left "Star (*) is not supported."
+            TH.ConstraintT -> Left "Constraints are not supported."
+            TH.LitT _ -> Left "Type-level literals are not supported."
 #if MIN_VERSION_template_haskell(2,11,0)
-            TH.InfixT{} -> fail "Infix type constructors are not supported."
-            TH.UInfixT{} -> fail "Unresolved infix type constructors are not supported."
-            TH.ParensT _ -> fail "Parenthesized types are not supported."
-            TH.WildCardT -> fail "Wildcard types are not supported."
+            TH.InfixT{} -> Left "Infix type constructors are not supported."
+            TH.UInfixT{} -> Left "Unresolved infix type constructors are not supported."
+            TH.ParensT _ -> Left "Parenthesized types are not supported."
+            TH.WildCardT -> Left "Wildcard types are not supported."
 #endif
 
 -- | Is the type a function?
@@ -340,11 +340,11 @@ reifyTypeDefinition typeConstructor@(TypeConstructor name) =
                    do ty' <- normalizeType ty
                       return (Just (TypeAliasDefinition typeConstructor
                                                         (TypeAlias (map toTypeVariable vars) ty')))
-                 _ -> fail "Not a supported data type declaration."
+                 _ -> Left "Not a supported data type declaration."
              TH.PrimTyConI{} -> return Nothing
-             TH.FamilyI{} -> fail "Data families not supported yet."
+             TH.FamilyI{} -> Left "Data families not supported yet."
              _ ->
-               fail ("Not a supported object, no type inside it: " ++
+               Left ("Not a supported object, no type inside it: " ++
                      TH.pprint info)
      case result of
        Left err -> fail err

--- a/src/Present.hs
+++ b/src/Present.hs
@@ -359,10 +359,18 @@ makeConstructor =
       ((\x y -> [x,y]) <$> makeSlot t1 <*> makeSlot t2)
     (TH.ForallC _ _ con) ->
       makeConstructor con
-    TH.GadtC _ _ _ ->
-      undefined -- FIXME
-    TH.RecGadtC _ _ _ ->
-      undefined -- FIXME
+    TH.GadtC names slots _type ->
+      case names of
+        name:_ -> -- FIXME: What about the other names?
+          Constructor <$> pure (ValueConstructor name) <*> mapM makeSlot slots
+        _ ->
+          Left "GADT constructors without a name are not supported."
+    TH.RecGadtC names fields _type ->
+      case names of
+        name:_ -> -- FIXME: What about the other names?
+          Constructor <$> pure (ValueConstructor name) <*> mapM makeField fields
+        _ ->
+          Left "GADT constructors without a name are not supported."
   where makeSlot (_,ty) = (Nothing,) <$> normalizeType ty
         makeField (name,_,ty) =
           (Just (ValueVariable name),) <$> normalizeType ty

--- a/src/Present.hs
+++ b/src/Present.hs
@@ -34,6 +34,7 @@ module Present
 
 import           Control.Arrow (second)
 import           Control.Exception (evaluate,SomeException(..),try,evaluate)
+import           Control.Monad (forM)
 import           Control.Monad.Trans.State.Strict (evalStateT,get,modify,StateT(..))
 import           Data.Char (isSpace,ord,isAlphaNum)
 import           Data.Int (Int8,Int16,Int32,Int64)
@@ -322,7 +323,7 @@ reifyTypeDefinition typeConstructor@(TypeConstructor name) =
 #else
                  TH.DataD _cxt _ vars cons _deriving ->
 #endif
-                   do cs <- mapM makeConstructor cons
+                   do cs <- concat <$> mapM makeConstructors cons
                       return (Just (DataTypeDefinition typeConstructor
                                                        (DataType (map toTypeVariable vars) cs)))
 #if MIN_VERSION_template_haskell(2,11,0)
@@ -330,11 +331,11 @@ reifyTypeDefinition typeConstructor@(TypeConstructor name) =
 #else
                  TH.NewtypeD _cxt _ vars con _deriving ->
 #endif
-                   do c <- makeConstructor con
+                   do cs <- makeConstructors con
                       return (Just (DataTypeDefinition
                                       typeConstructor
                                       (DataType (map toTypeVariable vars)
-                                                [c])))
+                                                cs)))
                  TH.TySynD _ vars ty ->
                    do ty' <- normalizeType ty
                       return (Just (TypeAliasDefinition typeConstructor
@@ -357,32 +358,29 @@ toTypeVariable =
     TH.KindedTV t _ -> TypeVariable t
 
 -- | Make a normalized constructor from the more complex TH Con.
-makeConstructor
-  :: TH.Con -> Either String Constructor
-makeConstructor =
+makeConstructors
+  :: TH.Con -> Either String [Constructor]
+makeConstructors =
   \case
-    TH.NormalC name slots ->
-      Constructor <$> pure (ValueConstructor name) <*> mapM makeSlot slots
-    TH.RecC name fields ->
-      Constructor <$> pure (ValueConstructor name) <*> mapM makeField fields
-    TH.InfixC t1 name t2 ->
-      Constructor <$> pure (ValueConstructor name) <*>
-      ((\x y -> [x,y]) <$> makeSlot t1 <*> makeSlot t2)
+    TH.NormalC name slots -> do
+      c <- Constructor <$> pure (ValueConstructor name) <*> mapM makeSlot slots
+      return [c]
+    TH.RecC name fields -> do
+      c <- Constructor <$> pure (ValueConstructor name) <*> mapM makeField fields
+      return [c]
+    TH.InfixC t1 name t2 -> do
+      c <- Constructor <$> pure (ValueConstructor name) <*>
+           ((\x y -> [x,y]) <$> makeSlot t1 <*> makeSlot t2)
+      return [c]
     (TH.ForallC _ _ con) ->
-      makeConstructor con
+      makeConstructors con
 #if MIN_VERSION_template_haskell(2,11,0)
     TH.GadtC names slots _type ->
-      case names of
-        name:_ -> -- FIXME: What about the other names?
-          Constructor <$> pure (ValueConstructor name) <*> mapM makeSlot slots
-        _ ->
-          Left "GADT constructors without a name are not supported."
+      forM names $ \name ->
+        Constructor <$> pure (ValueConstructor name) <*> mapM makeSlot slots
     TH.RecGadtC names fields _type ->
-      case names of
-        name:_ -> -- FIXME: What about the other names?
-          Constructor <$> pure (ValueConstructor name) <*> mapM makeField fields
-        _ ->
-          Left "GADT constructors without a name are not supported."
+      forM names $ \name ->
+        Constructor <$> pure (ValueConstructor name) <*> mapM makeField fields
 #endif
   where makeSlot (_,ty) = (Nothing,) <$> normalizeType ty
         makeField (name,_,ty) =

--- a/src/Present.hs
+++ b/src/Present.hs
@@ -1,3 +1,4 @@
+{-# LANGUAGE CPP #-}
 {-# LANGUAGE FlexibleContexts #-}
 {-# LANGUAGE OverloadedStrings #-}
 {-# LANGUAGE PatternGuards #-}
@@ -132,10 +133,12 @@ normalizeType = go
             TH.StarT -> fail "Star (*) is not supported."
             TH.ConstraintT -> fail "Constraints are not supported."
             TH.LitT _ -> fail "Type-level literals are not supported."
+#if MIN_VERSION_template_haskell(2,11,0)
             TH.InfixT{} -> fail "Infix type constructors are not supported."
             TH.UInfixT{} -> fail "Unresolved infix type constructors are not supported."
             TH.ParensT _ -> fail "Parenthesized types are not supported."
             TH.WildCardT -> fail "Wildcard types are not supported."
+#endif
 
 -- | Is the type a function?
 isFunction :: TH.Type -> Bool
@@ -314,11 +317,19 @@ reifyTypeDefinition typeConstructor@(TypeConstructor name) =
            case info of
              TH.TyConI dec ->
                case dec of
+#if MIN_VERSION_template_haskell(2,11,0)
                  TH.DataD _cxt0 _ vars _mkind cons _cxt1 ->
+#else
+                 TH.DataD _cxt _ vars cons _deriving ->
+#endif
                    do cs <- mapM makeConstructor cons
                       return (Just (DataTypeDefinition typeConstructor
                                                        (DataType (map toTypeVariable vars) cs)))
+#if MIN_VERSION_template_haskell(2,11,0)
                  TH.NewtypeD _cxt0 _ vars _mkind con _cxt1 ->
+#else
+                 TH.NewtypeD _cxt _ vars con _deriving ->
+#endif
                    do c <- makeConstructor con
                       return (Just (DataTypeDefinition
                                       typeConstructor
@@ -359,6 +370,7 @@ makeConstructor =
       ((\x y -> [x,y]) <$> makeSlot t1 <*> makeSlot t2)
     (TH.ForallC _ _ con) ->
       makeConstructor con
+#if MIN_VERSION_template_haskell(2,11,0)
     TH.GadtC names slots _type ->
       case names of
         name:_ -> -- FIXME: What about the other names?
@@ -371,6 +383,7 @@ makeConstructor =
           Constructor <$> pure (ValueConstructor name) <*> mapM makeField fields
         _ ->
           Left "GADT constructors without a name are not supported."
+#endif
   where makeSlot (_,ty) = (Nothing,) <$> normalizeType ty
         makeField (name,_,ty) =
           (Just (ValueVariable name),) <$> normalizeType ty
@@ -999,7 +1012,11 @@ getPresentInstances =
                TH.ClassI (TH.ClassD _ _ _ _ [TH.SigD method _]) instances ->
                  return (mapMaybe (\i ->
                                      case i of
+#if MIN_VERSION_template_haskell(2,11,0)
                                        TH.InstanceD _moverlap _ (TH.AppT (TH.ConT _className) (TH.ConT typeName)) _ ->
+#else
+                                       TH.InstanceD _ (TH.AppT (TH.ConT _className) (TH.ConT typeName)) _ ->
+#endif
                                          Just (TypeConstructor typeName
                                               ,ValueVariable method)
                                        _ -> Nothing)
@@ -1070,8 +1087,12 @@ presentName name =
   do result <- tryQ (TH.reify name)
      case result of
        Nothing -> fail "Name `it' isn't in scope."
-       Just (TH.VarI _ ty _) -> TH.appE (presentType (return ty))
-                                          (TH.varE name)
+#if MIN_VERSION_template_haskell(2,11,0)
+       Just (TH.VarI _ ty _) ->
+#else
+       Just (TH.VarI _ ty _ _) ->
+#endif
+         TH.appE (presentType (return ty)) (TH.varE name)
        _ -> fail "The name `it' isn't a variable."
   where tryQ m =
           TH.recover (pure Nothing)

--- a/src/Present.hs
+++ b/src/Present.hs
@@ -362,27 +362,24 @@ makeConstructors
   :: TH.Con -> Either String [Constructor]
 makeConstructors =
   \case
-    TH.NormalC name slots -> do
-      c <- Constructor <$> pure (ValueConstructor name) <*> mapM makeSlot slots
-      return [c]
-    TH.RecC name fields -> do
-      c <- Constructor <$> pure (ValueConstructor name) <*> mapM makeField fields
-      return [c]
-    TH.InfixC t1 name t2 -> do
-      c <- Constructor <$> pure (ValueConstructor name) <*>
-           ((\x y -> [x,y]) <$> makeSlot t1 <*> makeSlot t2)
-      return [c]
+    TH.NormalC name slots ->
+      (:[]) <$> makeConstructor name (mapM makeSlot slots)
+    TH.RecC name fields ->
+      (:[]) <$> makeConstructor name (mapM makeField fields)
+    TH.InfixC t1 name t2 ->
+      (:[]) <$> makeConstructor name ((\x y -> [x,y]) <$> makeSlot t1 <*> makeSlot t2)
     (TH.ForallC _ _ con) ->
       makeConstructors con
 #if MIN_VERSION_template_haskell(2,11,0)
     TH.GadtC names slots _type ->
       forM names $ \name ->
-        Constructor <$> pure (ValueConstructor name) <*> mapM makeSlot slots
+        makeConstructor name (mapM makeSlot slots)
     TH.RecGadtC names fields _type ->
       forM names $ \name ->
-        Constructor <$> pure (ValueConstructor name) <*> mapM makeField fields
+        makeConstructor name (mapM makeField fields)
 #endif
-  where makeSlot (_,ty) = (Nothing,) <$> normalizeType ty
+  where makeConstructor name efields = Constructor (ValueConstructor name) <$> efields
+        makeSlot (_,ty) = (Nothing,) <$> normalizeType ty
         makeField (name,_,ty) =
           (Just (ValueVariable name),) <$> normalizeType ty
 

--- a/src/Present.hs
+++ b/src/Present.hs
@@ -132,6 +132,10 @@ normalizeType = go
             TH.StarT -> fail "Star (*) is not supported."
             TH.ConstraintT -> fail "Constraints are not supported."
             TH.LitT _ -> fail "Type-level literals are not supported."
+            TH.InfixT{} -> fail "Infix type constructors are not supported."
+            TH.UInfixT{} -> fail "Unresolved infix type constructors are not supported."
+            TH.ParensT _ -> fail "Parenthesized types are not supported."
+            TH.WildCardT -> fail "Wildcard types are not supported."
 
 -- | Is the type a function?
 isFunction :: TH.Type -> Bool
@@ -310,11 +314,11 @@ reifyTypeDefinition typeConstructor@(TypeConstructor name) =
            case info of
              TH.TyConI dec ->
                case dec of
-                 TH.DataD _cxt _ vars cons _deriving ->
+                 TH.DataD _cxt0 _ vars _mkind cons _cxt1 ->
                    do cs <- mapM makeConstructor cons
                       return (Just (DataTypeDefinition typeConstructor
                                                        (DataType (map toTypeVariable vars) cs)))
-                 TH.NewtypeD _cxt _ vars con _deriving ->
+                 TH.NewtypeD _cxt0 _ vars _mkind con _cxt1 ->
                    do c <- makeConstructor con
                       return (Just (DataTypeDefinition
                                       typeConstructor
@@ -355,6 +359,10 @@ makeConstructor =
       ((\x y -> [x,y]) <$> makeSlot t1 <*> makeSlot t2)
     (TH.ForallC _ _ con) ->
       makeConstructor con
+    TH.GadtC _ _ _ ->
+      undefined -- FIXME
+    TH.RecGadtC _ _ _ ->
+      undefined -- FIXME
   where makeSlot (_,ty) = (Nothing,) <$> normalizeType ty
         makeField (name,_,ty) =
           (Just (ValueVariable name),) <$> normalizeType ty
@@ -983,7 +991,7 @@ getPresentInstances =
                TH.ClassI (TH.ClassD _ _ _ _ [TH.SigD method _]) instances ->
                  return (mapMaybe (\i ->
                                      case i of
-                                       TH.InstanceD _ (TH.AppT (TH.ConT _className) (TH.ConT typeName)) _ ->
+                                       TH.InstanceD _moverlap _ (TH.AppT (TH.ConT _className) (TH.ConT typeName)) _ ->
                                          Just (TypeConstructor typeName
                                               ,ValueVariable method)
                                        _ -> Nothing)
@@ -1054,7 +1062,7 @@ presentName name =
   do result <- tryQ (TH.reify name)
      case result of
        Nothing -> fail "Name `it' isn't in scope."
-       Just (TH.VarI _ ty _ _) -> TH.appE (presentType (return ty))
+       Just (TH.VarI _ ty _) -> TH.appE (presentType (return ty))
                                           (TH.varE name)
        _ -> fail "The name `it' isn't a variable."
   where tryQ m =

--- a/stack.yaml
+++ b/stack.yaml
@@ -1,6 +1,3 @@
-flags: {}
-extra-package-dbs: []
 packages:
 - '.'
-extra-deps: []
 resolver: lts-8.3

--- a/stack.yaml
+++ b/stack.yaml
@@ -1,35 +1,6 @@
-# For more information, see: http://docs.haskellstack.org/en/stable/yaml_configuration.html
-
-# Specifies the GHC version and set of packages available (e.g., lts-3.5, nightly-2015-09-21, ghc-7.10.2)
-resolver: lts-4.1
-
-# Local packages, usually specified by relative directory name
+flags: {}
+extra-package-dbs: []
 packages:
 - '.'
-
-# Packages to be pulled from upstream that are not in the resolver (e.g., acme-missiles-0.3)
 extra-deps: []
-
-# Override default flag values for local packages and extra-deps
-flags: {}
-
-# Extra package databases containing global packages
-extra-package-dbs: []
-
-# Control whether we use the GHC we find on the path
-# system-ghc: true
-
-# Require a specific version of stack, using version ranges
-# require-stack-version: -any # Default
-# require-stack-version: >= 1.0.0
-
-# Override the architecture used by stack, especially useful on Windows
-# arch: i386
-# arch: x86_64
-
-# Extra directories used by stack for building
-# extra-include-dirs: [/path/to/dir]
-# extra-lib-dirs: [/path/to/dir]
-
-# Allow a newer minor version of GHC than the snapshot specifies
-# compiler-check: newer-minor
+resolver: lts-8.3

